mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala

```diff
@@ -0,0 +1,633 @@
+      Vectors.dense(currMin)
+    }
+
+    /**
+     * L2 (Euclidian) norm of each dimension.
+     */
+    def normL2: Vector = {
+      require(requestedMetrics.contains(NormL2))
+      require(totalWeightSum > 0, s"Nothing has been added to this summarizer.")
+
+      val realMagnitude = Array.ofDim[Double](n)
+
+      var i = 0
+      val len = currM2.length
+      while (i < len) {
+        realMagnitude(i) = math.sqrt(currM2(i))
+        i += 1
+      }
+      Vectors.dense(realMagnitude)
+    }
+
+    /**
+     * L1 norm of each dimension.
+     */
+    def normL1: Vector = {
+      require(requestedMetrics.contains(NormL1))
+      require(totalWeightSum > 0, s"Nothing has been added to this summarizer.")
+
+      Vectors.dense(currL1)
+    }
+  }
+
+  private case class MetricsAggregate(
+      requestedMetrics: Seq[Metrics],
+      requestedComputeMetrics: Seq[ComputeMetrics],
+      featuresExpr: Expression,
+      weightExpr: Expression,
+      mutableAggBufferOffset: Int,
+      inputAggBufferOffset: Int)
+    extends TypedImperativeAggregate[SummarizerBuffer] {
+
+    override def eval(state: SummarizerBuffer): InternalRow = {
+      val metrics = requestedMetrics.map({
+        case Mean => UnsafeArrayData.fromPrimitiveArray(state.mean.toArray)
+        case Variance => UnsafeArrayData.fromPrimitiveArray(state.variance.toArray)
+        case Count => state.count
+        case NumNonZeros => UnsafeArrayData.fromPrimitiveArray(
+          state.numNonzeros.toArray.map(_.toLong))
+        case Max => UnsafeArrayData.fromPrimitiveArray(state.max.toArray)
+        case Min => UnsafeArrayData.fromPrimitiveArray(state.min.toArray)
+        case NormL2 => UnsafeArrayData.fromPrimitiveArray(state.normL2.toArray)
+        case NormL1 => UnsafeArrayData.fromPrimitiveArray(state.normL1.toArray)
+      })
+      InternalRow.apply(metrics: _*)
+    }
+
+    override def children: Seq[Expression] = featuresExpr :: weightExpr :: Nil
+
+    override def update(state: SummarizerBuffer, row: InternalRow): SummarizerBuffer = {
+      // val features = udt.deserialize(featuresExpr.eval(row))
+      val featuresDatum = featuresExpr.eval(row).asInstanceOf[InternalRow]
+
+      val isDense = featuresDatum.getByte(0) == 1
+
+      val indices: ArrayData = if (isDense) null else featuresDatum.getArray(2)
+      val values: ArrayData = featuresDatum.getArray(3)
+      val _size = if (isDense) values.numElements() else featuresDatum.getInt(1)
+
+      val features = new TraversableIndexedSeq {
+
+        override def foreachActive(f: (Int, Double) => Unit): Unit = {
+          var i = 0
+          if (isDense) {
+            while (i < _size) {
+              f(i, values.getDouble(i))
+              i += 1
+            }
+          } else {
+            while (i < indices.numElements()) {
+              f(indices.getInt(i), values.getDouble(i))
+              i += 1
+            }
+          }
+        }
+
+        override def size: Int = _size
+      }
+
```

Comment:
Here I do not use `VectorUDT.deserialize` but directly manipulate the `UnsafeArrayData` coming from `InternalRow`(In dataframe, it is `UnsafeRow` actually), it can avoid data copy. cc @liancheng @cloud-fan